src/lib.rs

```diff
@@ -1,334 +1,201 @@
 // Copyright 2019 PingCAP, Inc.
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// See the License for the specific language governing permissions and
-// limitations under the License.
 
 //! Utility functions for generating Rust code from protobuf specifications.
 //!
 //! These functions panic liberally, they are designed to be used from build
 //! scripts, not in production.
 
 #[cfg(feature = "prost-codec")]
-pub use crate::wrapper::GenOpt;
-use std::fmt::Debug;
-use std::fs::{self, File};
-use std::io::Write;
-use std::path::Path;
+mod wrapper;
+
+#[cfg(feature = "protobuf-codec")]
+mod protobuf_impl;
 
 #[cfg(feature = "prost-codec")]
-mod wrapper;
+mod prost_impl;
 
-/// Generate Rust files from proto files (`files`).
-pub fn generate_files<T: AsRef<Path> + Debug>(includes: &[T], files: &[T], out_dir: &str) {
-    if Path::new(&out_dir).exists() {
-        fs::remove_dir_all(&out_dir).unwrap();
-    }
-    fs::create_dir_all(&out_dir).unwrap();
-    generate(includes, files, out_dir);
-    let mut generated = Vec::new();
-    let modules: Vec<_> = fs::read_dir(out_dir)
-        .unwrap()
-        .filter_map(|res| {
-            let path = match res {
-                Ok(e) => e.path(),
-                Err(e) => panic!("failed to list {}: {:?}", out_dir, e),
-            };
-            if path.extension() == Some(std::ffi::OsStr::new("rs")) {
-                generated.push(format!("{}", path.display()));
-                let name = path.file_stem().unwrap().to_str().unwrap();
-                Some((name.replace('-', "_"), name.to_owned()))
-            } else {
-                None
-            }
-        })
-        .collect();
-    let mut f = File::create(format!("{}/mod.rs", out_dir)).unwrap();
-    for (module, file_name) in &modules {
-        if cfg!(feature = "protobuf-codec") {
-            writeln!(f, "pub mod {};", module).unwrap();
-            continue;
-        }
-        if module.starts_with("wrapper_") {
-            continue;
-        }
-        let mut level = 0;
-        for part in module.split('.') {
-            writeln!(f, "{:level$}pub mod {} {{", "", part, level = level).unwrap();
-            level += 1;
-        }
-        writeln!(
-            f,
-            "{:level$}include!(\"{}.rs\");",
-            "",
-            file_name,
-            level = level
+use bitflags::bitflags;
+use lazy_static::lazy_static;
+use std::fs::{self, File};
+use std::io::Write;
+use std::path::{Path, PathBuf};
+
+lazy_static! {
+    static ref OUT_DIR: String = {
```

Comment:
I think it's more flexible to make it a default value in builder instead of a static variables. For example, when building grpcio-protos, there are several modules to build, and better keep each modules in separate out_dir. Besides, static is bad.

Reply:
Sure can do. My reasoning was that we only want to generate into the target directory. If we accommodate generating into source, then we might end up supporting formatting, etc.